effectful/src/Effectful/STE.hs

```diff
@@ -0,0 +1,84 @@
+module Effectful.STE
+  ( STE
+
+  , MutVar
+  , runSTE
+  , newMutVar
+  , readMutVar
+  , writeMutVar
+  , atomicModifyMutVar
+  , modifyMutVar
+
+  , MutVarU
+  , newMutVarU
+  , readMutVarU
+  , writeMutVarU
+  , modifyMutVarU
+  ) where
+
+import Control.Monad.Primitive
+import Data.Coerce
+import Data.Primitive.Types
+import qualified Data.Primitive.MutVar as P
+import qualified Data.Primitive.PrimArray as P
+
+import Effectful
+import Effectful.Dispatch.Static
+import Effectful.Internal.Effect
+
+-- | Dirty trick for disambiguation of STE type parameters.
+instance {-# INCOHERENT #-} s1 ~ s2 => STE s1 :> (STE s2 : es) where
+  reifyIndex = 0
+
+type role STE nominal phantom phantom
```

Comment:
I'm not sure what is appropriate role for the last two type arguments, but the `s` is absolutely phantom. How is `nominal` appropriate?

Reply:
This just mimics the API of `ST`, where `s` is `nominal`. I don't know if making it phantom has any consequences, so I left it like that.

For sure `MutVar` etc. need the `s` to be `nominal`, otherwise you can just cast it to `()` with `coerce` and pass it outside of the scope of `runSTE`.

Reply:
Indeed! Thanks.